src/ucs/sys/topo/base/topo.h

```diff
@@ -53,9 +54,52 @@ typedef struct ucs_sys_dev_distance {
 } ucs_sys_dev_distance_t;
 
 
+/*
+ * Captures common entity between two system devices to allow for decisions
+ * based on system configuration instead of specific bandwidth/latency
+ * that ucs_sys_dev_distance_t provides.
+ */
+typedef enum ucs_sys_common_ancestor {
+    UCS_SYS_TOPO_COMMON_DEVICE,
+    UCS_SYS_TOPO_COMMON_PCIE_SWITCH,
+    UCS_SYS_TOPO_COMMON_PCIE_SWITCH_HIERARCHY,
+    UCS_SYS_TOPO_COMMON_HOST_BRIDGE,
+    UCS_SYS_TOPO_COMMON_NUMA_NODE,
+    UCS_SYS_TOPO_COMMON_SYSTEM,
+    UCS_SYS_TOPO_COMMON_UNKNOWN
+} ucs_sys_common_ancestor_t;
```

Comment:
not sure we need it - how can we use such info in UCP?
in any case i think it should be something more abstract

Reply:
@yosefe The aim was to have another decision process in rendezvous protocols. If we only have get_distance, then we would need to make decisions along the lines of 
```
 ucs_topo_get_distance(dev1, dev2, &distance);
  if ((system == X) && (distance.bw > low_thresh) && (distance.bw< high_thresh)) {
  } else if (...) {
    ...
  } ...
```

But adding common_ancestor allows for system-agnostic decisions: 
```
 ucs_topo_get_common_ancestor(dev1, dev2, &common_ancestor);
  if (common_ancestory = PCIE_SWITCH) {
  } else if (...) {
    ...
  } ...
```

Reply:
why need system==X in the first snippet?

Reply:
I would think that low_thresh/high_thresh would be system X dependent. I guess the following would be a better way to put it:
```
 ucs_topo_get_distance(dev1, dev2, &distance);
  if ((distance.bw > low_thresh(X)) && (distance.bw< high_thresh(X))) {
  } else if (...) {
    ...
  } ...
```

Reply:
1. I'd prefer to design topo.h API to be system-agnostic and in generic terms like distance or even graphs and not tie it to a particular HW architecture
2. If we manage system-specific branches to ucp.h it would become a mess of obscure behaviors. need abstract approach on top of topo.h
3. which optimizations or tunings cannot be expressed by lat/bw distance as we have today?

Reply:
Agree with 1. and 2.

For 3., if we want to make a decision of whether to use device bounce buffers for rendezvous pipeline, we'd have to check if local device buffers are accessible device bounce buffers of remote process. Assuming we pack bus id (which would be valid within the node) of remote device buffers into RTS, in order to check if remote device bounce buffers are reachable, the local process may call ucs_topo_get_distance and get back a result (let's say 25 GB/s). On DGX-1 machine with v100s, this would mean there are nvlinks between the two devices; but on a system with A100s this would likely be PCIe gen4 bandwidth and indicate absence of nvlinks between the two devices and we may prefer to not use device-staged pipeline. As we don't have a way to estimate latency accurately between devices using nvml, it may not help with the protocol decision and it may be the case that decisions have to system-specific breaking requirement 2).  

I was hoping a function like get_common_ancestor would be better in being system-agnostic. That said, I don't disagree that switch, switch_hierarchy, host_bridge, etc are not abstract enough and too hw-specific that UCP should have to deal with. Is hop count a better abstraction? topology modules can provide a reasonable estimate of that if not the exact. In this case, we could decide to use device-staged pipeline if remote device is exactly 1 hop away. Maybe this is what you meant by graphs?

Reply:
@yosefe I've removed common_ancestor functions from this PR in order to keep maintain the original scope. Maybe we'll bring this back in a separate PR.